cesium/internal/domain/delete.go

```diff
@@ -130,107 +62,224 @@ func (db *DB) Delete(ctx context.Context, startPosition int, endPosition int, st
+	)
+
+	// Find all pointers using the file: there cannot be more pointers using the file
+	// during GC since the file must be already full – however, there can be less due
+	// to deletion.
+	db.idx.mu.RLock()
+	for _, ptr := range db.idx.mu.pointers {
+		if ptr.fileKey == key {
+			ptrs = append(ptrs, ptr)
+			tombstoneSize -= int64(ptr.length)
 		}
+	}
+	db.idx.mu.RUnlock()
+
+	// Decide whether we should GC
+	if tombstoneSize < int64(db.GCThreshold*float32(db.FileSize)) {
+		return nil
+	}
+
+	// Open a reader on the old file.
+	r, err := db.FS.Open(name, os.O_RDONLY)
+	if err != nil {
+		return err
+	}
+
+	// Open a writer to the copy file.
+	w, err := db.FS.Open(copyName, os.O_WRONLY|os.O_CREATE)
+	if err != nil {
+		return err
+	}
 
-		err = db.files.removeReadersWriters(ctx, fileKey)
+	// Find all pointers stored in the old file, and write them to the new file.
+	for _, ptr := range ptrs {
+		buf := make([]byte, ptr.length)
+		_, err = r.ReadAt(buf, int64(ptr.offset))
 		if err != nil {
 			return err
 		}
 
-		err = db.FS.Remove(strconv.Itoa(int(fileKey)) + ".domain")
+		_, err = w.WriteAt(buf, int64(newOffset))
 		if err != nil {
 			return err
 		}
-		err = db.FS.Rename(strconv.Itoa(int(fileKey))+"_temp.domain", strconv.Itoa(int(fileKey))+".domain")
-		if err != nil {
-			return err
+
+		if newOffset != ptr.offset {
+			offsetDeltaMap[ptr.TimeRange] = ptr.offset - newOffset
+		}
+		newOffset += ptr.length
+	}
+
+	if err = r.Close(); err != nil {
+		return err
+	}
+
+	if err = w.Close(); err != nil {
+		return err
+	}
+
+	// Update the file and index while holding the mutex lock.
+	// Note: the index might be different at this point than before: old pointers on
+	// this file may be split into multiple smaller pointers with different offsets.
+	// (Understand that since this deletion occurred after garbage collection, it should
+	// not be garbage collected in this run of GC.)
+	// However, two things cannot change:
+	// 1. The resulting pointers from a pointer deletion, no matter into how many,
+	// cannot end up in a larger timerange than the original pointer.
+	// 2. The delta in offset, i.e. oldOffset - newOffset are the same for all smaller,
+	// resulting pointers from the original split.
+	//
+	// Using these two principles, we can find the new offset for any pointer with a
+	// timerange contained in the original pointer by subtracting it by the same delta.
+	db.idx.mu.Lock()
+	for i, ptr := range db.idx.mu.pointers {
+		if ptr.fileKey == key {
+			deltaOffset, ok := resolvePointerOffset(ptr.TimeRange, offsetDeltaMap)
+			if ok {
+				db.idx.mu.pointers[i].offset = ptr.offset - deltaOffset
+			}
+		}
+	}
+
+	err = db.FS.Rename(name, name+"_temp")
+	if err != nil {
+		db.idx.mu.Unlock()
+		return err
+	}
+	err = db.FS.Rename(copyName, name)
+	if err != nil {
+		db.idx.mu.Unlock()
+		return err
+	}
+	db.idx.mu.Unlock()
+
+	if err = db.files.rejuvenate(key); err != nil {
+		return err
+	}
+
+	return db.FS.Remove(name + "_temp")
```

Comment:
Let's say the garbage collection fails halfway through. We probably want to try to recover the DB to a certain state no? This probably means renaming files again and removing certain things. Stuff to think about